java/Ice/context/server/src/main/java/com/zeroc/demos/greeter/Server.java

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) ZeroC, Inc.
+
+package com.zeroc.demos.greeter;
+
+import com.zeroc.Ice.Communicator;
+import com.zeroc.Ice.ObjectAdapter;
+
+class Server {
+    public static void main(String[] args) {
+        // Create an Ice communicator. We'll use this communicator to create an object adapter.
+        try (Communicator communicator = com.zeroc.Ice.Util.initialize(args)) {
+
+            // Register a shutdown hook that calls communicator.shutdown() when the user shuts down the server with
+            // Ctrl+C or similar. The shutdown hook thread also waits until the main thread completes its cleanup.
+            shutdownCommunicatorOnCtrlC(communicator, Thread.currentThread());
+
+            // Create an object adapter that listens for incoming requests and dispatches them to servants.
+            ObjectAdapter adapter = communicator.createObjectAdapterWithEndpoints("GreeterAdapter", "tcp -p 4061");
+
+            // Register the Chatbot servant with the adapter.
+            adapter.add(new Chatbot(), com.zeroc.Ice.Util.stringToIdentity("greeter"));
+
+            // Start dispatching requests.
+            adapter.activate();
+            System.out.println("Listening on port 4061...");
+
+            // Wait until the communicator is shut down. Here, this occurs when the user presses Ctrl+C.
+            communicator.waitForShutdown();
+        }
+    }
+
+    private static void shutdownCommunicatorOnCtrlC(Communicator communicator, Thread mainThread) {
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            System.out.println("Caught Ctrl+C, shutting down...");
+            communicator.shutdown();
+
+            // Wait until the main thread completes.
+            try {
+                mainThread.join();
+            } catch (InterruptedException e) {
+                assert false : "Shutdown thread cannot be interrupted";
```

Comment:
Is this saying that it's impossible to get here as a shutdown thread cannot be interrupted, or that the user should not try to interrupted a shutdown thread?

Reply:
We create a new Thread in the addShutdownHook call. The application doesn't keep a reference to this thread and nowhere does it attempt to interrupt it.

Maybe we should have instead:
```java
assert false; // this can't happen
```

We have to catch InterruptedException because it's a checked exception.